cpp/src/io/json/json_column.cu

```diff
@@ -593,29 +638,36 @@ void make_device_json_column(device_span<SymbolT const> input,
     auto& parent_col = it->second.get();
     bool replaced    = false;
     if (mapped_columns.count({parent_col_id, name}) > 0) {
+      auto const old_col_id = mapped_columns[{parent_col_id, name}];
       // If mixed type as string is enabled, make both of them strings and merge them.
       // All child columns will be ignored when parsing.
       if (is_enabled_mixed_types_as_string) {
-        // VAL/STR or STRUCT or LIST
-        auto old_col_id = mapped_columns[{parent_col_id, name}];
-
-        is_mixed_type_column[this_col_id] = 1;
-        is_mixed_type_column[old_col_id]  = 1;
-        // if old col type (not cat) is list or struct, replace with string.
-        auto& col = columns.at(old_col_id).get();
-        if (col.type == json_col_t::ListColumn or col.type == json_col_t::StructColumn) {
-          reinitialize_as_string(old_col_id, col);
-          // all its children (which are already inserted) are ignored later.
+        bool is_mixed_type = true;
+        // If new or old is STR and they are all not null, make it mixed type, else ignore.
+        if (column_categories[this_col_id] == NC_VAL || column_categories[this_col_id] == NC_STR) {
+          if (is_str_column_all_nulls[this_col_id]) is_mixed_type = false;
+        }
+        if (column_categories[old_col_id] == NC_VAL || column_categories[old_col_id] == NC_STR) {
+          if (is_str_column_all_nulls[old_col_id]) is_mixed_type = false;
+        }
```

Comment:
Could wrap this in a lambda and declare `is_mixed_type` as a `bool const`

Reply:
updated.